feat(sidebar): added sidebar component

mcoker opened the pull request:
fixes https://github.com/patternfly/patternfly/issues/3653

This adds a basic sidebar component. In a follow up sprint, I would like to add responsive support to the stack layout, split layout, gutters, sticky, and right panel modifiers.

patternfly-build commented:
Preview: https://patternfly-pr-3801.surge.sh

mcarrano reviewed (commented):
Looks good @mcoker , but I have a few questions.

* What's the purpose of the Stacked example? Seems counter-intuitive to have a stacked siderbar.

* It would be great if you can either add some documentation or something visual that differentiates your examples better. It's hard for me to tell what's different about "basic," "gutter," and "split" for example.

* Why do we have 2 sticky panel examples? Are they intended to be different?

DaoDaoNoCode reviewed (commented):
@mcoker Looks great! And I agree with @mcarrano . I cannot tell the difference between basic and split sidebar, and I can see there is a space between panel and content for gutter sidebar, but it's not obvious. Is there a way to make it more obvious? Like adding background/border/divider to the example?

DaoDaoNoCode commented on src/patternfly/components/Sidebar/examples/Sidebar.md:
```diff
@@ -0,0 +1,171 @@
+      <p>Moves the panel to the right by placing the panel after the content in the HTML.</p>
+      <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse dapibus nulla id augue dictum commodo. Donec mollis arcu massa, sollicitudin venenatis est rutrum vitae. Integer pulvinar ligula at augue mollis, ac pulvinar arcu semper. Maecenas nisi lorem, malesuada ac lectus nec, porta pretium neque. Ut convallis libero sit amet metus mattis, vel facilisis lorem malesuada. Duis cons ...
+      <p>Praesent at odio nec sapien ultrices tincidunt in non mauris. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Duis consectetur nisl quis facilisis faucibus. Sed eu bibendum risus. Suspendisse porta euismod tortor, at elementum odio suscipit sed. Cras eget ultrices urna, ac feugiat lectus. Integer a pharetra velit, in imperdiet mi. Phasellus vel hend ...
+    {{/content}}
```
Do we need to place the content before panel in this example?

mcoker replied:
Yep! Updated.

DaoDaoNoCode commented on src/patternfly/components/Sidebar/examples/Sidebar.md:
```diff
@@ -0,0 +1,171 @@
+{{#> sidebar sidebar--modifier="pf-m-stack"}}
+  {{#> sidebar-panel}}
+    Sidebar panel
+  {{/sidebar-panel}}
```
The panel here seems not aligned to the top of `__main`, is it designed to do so?

mcoker replied:
The way sticky works, it's not really designed to do that, it's designed to stick to the top of the closest ancestor with `overflow` set to scroll. I moved the local styles that define height and overflow to another workspace example element though, hopefully that behaves more like we would expect.

mcoker commented:
>  I can see there is a space between panel and content for gutter sidebar, but it's not obvious. Is there a way to make it more obvious? Like adding background/border/divider to the example?

Updated the example so there is more content in the panel and the gutter is more visible.

mcoker commented:
> What's the purpose of the Stacked example?

It's basically the responsive version. By default the layout is stacked up to the medium breakpoint, then it changes to a side by side layout. This allows the user to specify the layout is stacked by default, then split at the XL breakpoint. Or split by default, stacked at medium, split at XL, etc.

> It would be great if you can either add some documentation or something visual that differentiates your examples better.

Added descriptions in the example.

> Why do we have 2 sticky panel examples? Are they intended to be different?

Nope! Copy 'n paste error.

mcarrano commented:
Thanks for the clarifications @mcoker

mattnolting commented on src/patternfly/components/Sidebar/examples/Sidebar.md:
```diff
@@ -0,0 +1,171 @@
+| Class | Applied to | Outcome |
+| -- | -- | -- |
+| `.pf-c-sidebar` | `<div>` | Initiates the sidebar component. **Required** |
+| `.pf-c-simple-list__main` | `<div>` | Initiates the sidebar main element. **Required** |
+| `.pf-c-simple-list__panel` | `<div>` | Initiates the sidebar panel element. **Required** |
+| `.pf-c-simple-list__content` | `<div>` | Initiates the sidebar content element. **Required** |
```
```suggestion
| `.pf-c-sidebar__main` | `<div>` | Initiates the sidebar main element. **Required** |
| `.pf-c-sidebar__panel` | `<div>` | Initiates the sidebar panel element. **Required** |
| `.pf-c-sidebar__content` | `<div>` | Initiates the sidebar content element. **Required** |
```

mattnolting commented on src/patternfly/components/Sidebar/sidebar.scss:
```diff
@@ -0,0 +1,184 @@
+  --pf-c-sidebar--BackgroundColor: var(--pf-global--BackgroundColor--100);
+
+  // Main
+  --pf-c-sidebar__main--FlexDirection: column;
+  --pf-c-sidebar__main--md--FlexDirection: row;
```
👍

mattnolting commented on src/patternfly/components/Sidebar/sidebar-content.hbs:
```diff
@@ -0,0 +1,6 @@
+  {{#if sidebar-content--attribute}}
+    {{{sidebar-content--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
```suggestion
  {{> @partial-block}}
```

mattnolting commented on src/patternfly/components/Sidebar/sidebar-main.hbs:
```diff
@@ -0,0 +1,6 @@
+  {{#if sidebar-main--attribute}}
+    {{{sidebar-main--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
```suggestion
  {{> @partial-block}}
```

mattnolting commented on src/patternfly/components/Sidebar/sidebar-panel.hbs:
```diff
@@ -0,0 +1,6 @@
+  {{#if sidebar-panel--attribute}}
+    {{{sidebar-panel--attribute}}}
+  {{/if}}>
+  {{>@partial-block}}
```
```suggestion
  {{> @partial-block}}
```

mattnolting reviewed (changes requested):
This looks great! Super simple and flexible. Just a few housekeeping notes.

mcoker commented:
@DaoDaoNoCode @mattnolting updated!

mattnolting reviewed (approved):
LPTM 🎉

DaoDaoNoCode reviewed (approved):
Looks great to me, Thank you @mcoker !

mcarrano reviewed (approved):
Looks good. Please go ahead and merge.

redallen commented:
:tada: This PR is included in version 4.79.0 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v4.79.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/4.79.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
